planner/core/rule_aggregation_elimination.go

```diff
@@ -155,3 +272,174 @@ func (a *aggregationEliminator) optimize(p LogicalPlan) (LogicalPlan, error) {
 	}
 	return p, nil
 }
+
+func collectNestedAggPattern(lp LogicalPlan, ptn *nestedAggPattern) {
+	if len(lp.Children()) != 1 {
+		return
+	}
+	child := lp.Children()[0]
+	switch v := child.(type) {
+	case *LogicalAggregation:
+		ptn.inner = v
+		return
+	case *LogicalProjection:
+		if ptn.proj != nil {
+			return
+		}
+		ptn.proj = v
+	case *LogicalSort:
+		// Ignore sort between aggregations.
+	default:
+		ptn.isTrivial = false
+	}
+	collectNestedAggPattern(child, ptn)
+}
+
+// tryToCombineAggFunc checks the types of inner/outer aggregate function and check if they can be combined as one based on their semantics.
+// for example, since max(max(PARTIAL)) can be combined as max(TOTAL), we can combine inner max() and outer max() as a final max()
+// `innerIsGroup` indicates if the inner aggregate function is aggregating group-by items(values are de-duplicated).
+func tryToCombineAggFunc(outer, inner *aggregation.AggFuncDesc, innerIsGroup bool, expr expression.Expression) *aggregation.AggFuncDesc {
+	combined := outer.Clone()
+	combined.Args[0] = expr
+
+	if innerIsGroup {
+		if (inner.Name == ast.AggFuncFirstRow || inner.Name == ast.AggFuncMax || inner.Name == ast.AggFuncMin) &&
+			(outer.Name == ast.AggFuncFirstRow || outer.Name == ast.AggFuncMax || outer.Name == ast.AggFuncMin) {
+			return combined
+		}
+		// TODO: add type cast to support more cases
+	}
+
+	switch {
+	case inner.Name == ast.AggFuncFirstRow && outer.Name == ast.AggFuncFirstRow:
+	case inner.Name == ast.AggFuncMax && outer.Name == ast.AggFuncMax:
+	case inner.Name == ast.AggFuncMin && outer.Name == ast.AggFuncMin:
+	case inner.Name == ast.AggFuncBitAnd && outer.Name == ast.AggFuncBitAnd:
+	case inner.Name == ast.AggFuncBitOr && outer.Name == ast.AggFuncBitOr:
+	case inner.Name == ast.AggFuncBitXor && outer.Name == ast.AggFuncBitXor:
+	case inner.Name == ast.AggFuncGroupConcat && outer.Name == ast.AggFuncGroupConcat:
+		if inner.HasDistinct {
+			return nil
+		}
+	case inner.Name == ast.AggFuncSum && outer.Name == ast.AggFuncSum:
+		if inner.HasDistinct || outer.HasDistinct {
+			return nil
+		}
+	case inner.Name == ast.AggFuncCount && outer.Name == ast.AggFuncSum:
+		// TODO: add cast to support `sum(count(_)) -> count(_)`
+		return nil
+	default:
+		return nil
+	}
```

Comment:
`count(count(_))` can be transformed to `count(_)` as well, and `first_row / min / max` is applicable even if `HasDistinct` is true. This check looks messy, can we make it consistent with `isDecomposable` currently? From my understanding, generally we can safely merge aggregations if they share same name and the aggregation is decomposable, then we consider special cases like `sum(count(_)) -> count(_)` where the aggregations are different.